packages/agreement-platformstate-writer/src/consumerServiceV2.ts

```diff
@@ -1,27 +1,281 @@
+        );
+
+        if (!agreementEntry || agreementEntry.version > msg.version) {
+          return Promise.resolve();
+        } else {
+          await updateAgreementStateInPlatformStatesEntry(
+            dynamoDBClient,
+            primaryKey,
+            agreementStateToItemState(agreement.state),
+            msg.version
+          );
+
+          const GSIPK_consumerId_eserviceId = makeGSIPKConsumerIdEServiceId({
+            consumerId: agreement.consumerId,
+            eserviceId: agreement.eserviceId,
+          });
+
+          if (
+            await isAgreementTheLatest(
+              GSIPK_consumerId_eserviceId,
+              agreement.id,
+              dynamoDBClient
+            )
+          ) {
+            // token-generation-states only if agreement is the latest
+
+            await updateAgreementStateInTokenGenerationStatesTable({
+              GSIPK_consumerId_eserviceId,
+              agreementState: agreement.state,
+              dynamoDBClient,
+            });
+          }
+        }
+      }
+    )
+    .with({ type: "AgreementUpgraded" }, async (msg) => {
+      const agreement = parseAgreement(msg.data.agreement);
+      const primaryKey = makePlatformStatesAgreementPK(agreement.id);
+      const agreementEntry = await readAgreementEntry(
+        primaryKey,
+        dynamoDBClient
+      );
+
+      const pkCatalogEntry = makePlatformStatesEServiceDescriptorPK({
+        eserviceId: agreement.eserviceId,
+        descriptorId: agreement.descriptorId,
+      });
+      const catalogEntry = await readCatalogEntry(
+        pkCatalogEntry,
+        dynamoDBClient
+      );
+      if (!catalogEntry) {
+        // TODO double-check
+        throw genericInternalError("Catalog entry not found");
+      }
+
+      const GSIPK_consumerId_eserviceId = makeGSIPKConsumerIdEServiceId({
+        consumerId: agreement.consumerId,
+        eserviceId: agreement.eserviceId,
+      });
+
+      if (agreementEntry) {
+        if (agreementEntry.version > msg.version) {
+          return Promise.resolve();
+        } else {
+          await updateAgreementStateInPlatformStatesEntry(
+            dynamoDBClient,
+            primaryKey,
+            agreementStateToItemState(agreement.state),
+            msg.version
+          );
+        }
+      } else {
+        const newAgreementEntry: PlatformStatesAgreementEntry = {
+          PK: primaryKey,
+          state: agreementStateToItemState(agreement.state),
+          version: msg.version,
+          updatedAt: new Date().toISOString(),
+          GSIPK_consumerId_eserviceId,
+          GSISK_agreementTimestamp:
+            // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+            agreement.stamps.activation!.when.toISOString(),
+          agreementDescriptorId: agreement.descriptorId,
+        };
+
+        await writeAgreementEntry(newAgreementEntry, dynamoDBClient);
+      }
+
+      const doOperationOnTokenStates = async (
```

Comment:
Renamed to `updateLatestAgreementOnTokenStates`

I don't like it to be here, too. However here it's easier to review since it's just where it is called.